dll/win32/shell32/folders/CControlPanelFolder.cpp

```diff
@@ -773,3 +773,144 @@ HRESULT WINAPI CCPLItemMenu::HandleMenuMsg(
 
     return E_NOTIMPL;
 }
+
+/**************************************************************************
+* COpenControlPanel
+*/
+
+static HRESULT GetParsingName(PCIDLIST_ABSOLUTE pidl, PWSTR*Name)
+{
+    PIDLIST_ABSOLUTE pidlFree = NULL;
+    if (IS_INTRESOURCE(pidl))
+    {
+        HRESULT hr = SHGetSpecialFolderLocation(NULL, (UINT)(SIZE_T)pidl, &pidlFree);
+        if (FAILED(hr))
+            return hr;
+        pidl = pidlFree;
+    }
+    HRESULT hr = SHGetNameFromIDList(pidl, SIGDN_DESKTOPABSOLUTEPARSING, Name);
+    ILFree(pidlFree);
+    return hr;
+}
+
+static HRESULT CreateCplAbsoluteParsingPath(LPCWSTR Prefix, LPCWSTR InFolderParse, PWSTR Buf, UINT cchBuf)
+{
+    PWSTR cpfolder;
+    HRESULT hr = GetParsingName((PCIDLIST_ABSOLUTE)CSIDL_CONTROLS, &cpfolder);
+    if (SUCCEEDED(hr))
+    {
+        hr = StringCchPrintfW(Buf, cchBuf, L"%s\\%s%s", cpfolder, Prefix, InFolderParse);
+        SHFree(cpfolder);
+    }
+    return hr;
+}
+
+static HRESULT FindExeCplClass(LPCWSTR Canonical, HKEY hKey, BOOL Wow64, LPWSTR clsid)
+{
+    HRESULT hr = HRESULT_FROM_WIN32(ERROR_NOT_FOUND);
+    HKEY hNSKey;
+    WCHAR key[MAX_PATH], buf[MAX_PATH];
+    wsprintfW(key, L"Software\\Microsoft\\Windows\\CurrentVersion\\Explorer\\%s\\NameSpace",
+              Wow64 ? L"ControlPanelWOW64" : L"ControlPanel");
+    LSTATUS error = RegOpenKeyExW(hKey, key, 0, KEY_READ, &hNSKey);
+    if (error)
+        return HRESULT_FROM_WIN32(error);
+    for (DWORD i = 0; RegEnumKeyW(hNSKey, i, key, _countof(key)) == ERROR_SUCCESS; ++i)
+    {
+        IID validate;
+        if (SUCCEEDED(IIDFromString(key, &validate)))
+        {
+            wsprintfW(buf, L"CLSID\\%s", key);
+            DWORD cb = sizeof(buf);
+            if (RegGetValueW(HKEY_CLASSES_ROOT, buf, L"System.ApplicationName",
+                             RRF_RT_REG_SZ, NULL, buf, &cb) == ERROR_SUCCESS)
+            {
+                if (!lstrcmpiW(buf, Canonical))
+                {
+                    lstrcpyW(clsid, key);
+                    hr = S_OK;
+                }
+            }
+        }
+    }
+    RegCloseKey(hNSKey);
+    return hr;
+}
+
+static HRESULT FindExeCplClass(LPCWSTR Canonical, LPWSTR clsid)
+{
+    HRESULT hr = E_FAIL;
+    if (FAILED(hr))
+        hr = FindExeCplClass(Canonical, HKEY_CURRENT_USER, FALSE, clsid);
+    if (FAILED(hr))
+        hr = FindExeCplClass(Canonical, HKEY_CURRENT_USER, TRUE, clsid);
```

Comment:
Shouldn't the TRUE cases (WoW64) be called only if `IsWow64Process()` returns TRUE?

Reply:
I don't see the harm in checking both. The name is `companyname.whatever`, it is not going to clash with anything else.

When actual wow64 registry support lands, feel free to nag me for a fix if you find a case that actually breaks anything.